gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/activity/impl/DeleteWorkDirsActivityImpl.java

```diff
@@ -0,0 +1,224 @@
+          cleanupStagingDataPerTask(jobState, workSpec) : cleanupStagingDataForEntireJob(jobState, resourcesToClean, workSpec);
+
+      return new CleanupResult(cleanupMetrics);
+    } catch (Exception e) {
+      throw ApplicationFailure.newNonRetryableFailureWithCause(
+          String.format("Failed to cleanup temporary folders for job %s", optJobName.orElse(UNDEFINED_JOB_NAME)),
+          IOException.class.toString(),
+          new IOException(e)
+      );
+    }
+  }
+
+  private boolean canCleanStagingData(JobContext jobContext, JobState jobState)
+      throws IOException {
+    return jobContext.getSemantics() != DeliverySemantics.EXACTLY_ONCE || !jobContext.getCommitSequenceStore()
+        .get().exists(jobState.getJobName());
+  }
+
+  // Goes through all the task states and cleans them up individually, there is some overlap with the job level cleanup as the job level cleanup will
+  // also clean up the task level staging data
+  private static Map<String, Long> cleanupStagingDataPerTask(JobState jobState, WUProcessingSpec workSpec) throws IOException {
+    Closer closer = Closer.create();
+    Map<String, ParallelRunner> parallelRunners = Maps.newHashMap();
+    FileSystem fs = Help.loadFileSystem(workSpec);
+    int numThreads = jobState.getPropAsInt(ParallelRunner.PARALLEL_RUNNER_THREADS_KEY, ParallelRunner.DEFAULT_PARALLEL_RUNNER_THREADS);
+    List<TaskState> taskStateList = Help.loadTaskStates(workSpec, fs, jobState, numThreads);
+    Map<String, Long> cleanupMetrics = Maps.newHashMap();
+    try {
+      for (TaskState taskState : taskStateList) {
+        try {
+          cleanupMetrics.putAll(cleanTaskStagingData(taskState, log, closer, parallelRunners));
+        } catch (IOException e) {
+          log.error(String.format("Failed to clean staging data for task %s: %s", taskState.getTaskId(), e), e);
+        }
+      }
+    } finally {
+      try {
+        closer.close();
+      } catch (IOException e) {
+        log.error("Failed to clean staging data", e);
+      }
+    }
+    return cleanupMetrics;
+  }
+
+  private static Map<String, Long> cleanupStagingDataForEntireJob(JobState state, Set<String> resourcesToClean, WUProcessingSpec workSpec) throws IOException {
+    if (!state.contains(ConfigurationKeys.WRITER_STAGING_DIR) || !state.contains(ConfigurationKeys.WRITER_OUTPUT_DIR)) {
+      return Maps.newHashMap();
+    }
+    String writerFsUri = state.getProp(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, ConfigurationKeys.LOCAL_FS_URI);
+    FileSystem fs = JobLauncherUtils.getFsWithProxy(state, writerFsUri, WriterUtils.getFsConfiguration(state));
+    Map<String, Long> cleanupMetrics = Maps.newHashMap();
+    String jobId = state.getJobId();
+
+    for (String resource : resourcesToClean) {
+      Path pathToClean = new Path(resource);
+      Long filesCleaned = fs.getContentSummary(pathToClean).getLength();
```

Comment:
Gien recent issues wanted to see if there was an easy way through the return types to detect if a job deleted more than it was intended to...but is this overkill for this impl?

Reply:
I feel it might be overkill.  as a start, you could always log this value, but I'm not expecting you to always do that. 

(anyway, `.getLength()` should be the size in bytes, NOT `filesCleaned`.)